File: chain-of-responsibility-pattern/logger/logging-system.ts
abstract class Logger {
  private logLevel: string;
  private nextHandler: Logger | null;

  constructor(logLevel: string) {
    this.logLevel = logLevel;
    this.nextHandler = null;
  }

  /**
   * Sets the next handler in the chain of responsibility.
   *
   * @param handler - The next logger handler in the chain
   * @returns The next logger handler, allowing method chaining.
   */
  setNext(handler: Logger): Logger {
    this.nextHandler = handler;
    return handler;
  }

  /**
   * Handles the log message if level matches.
   *
   * If the level doesn't match, it passes the message to the next handler in the chain.
   *
   * @param level - The log level to handle (eg: INFO, DEBUG, etc)
   * @param message - The message to log.
   */
  handle(level: string, message: string): void {
    if (this.logLevel === level) {
      this.logMessage(message);
    } else if (this.nextHandler) {
      console.log("Match failed, passing it along", this.logLevel, level);
      this.nextHandler.handle(level, message);
    }
  }

  // This logic needs to be implemented by the child classes of this base class
  protected abstract logMessage(msg: string): void;
}

class InfoLogger extends Logger {
  protected logMessage(msg: string): void {
    console.info(`[INFO]: ${msg}`);
  }
}

class DebugLogger extends Logger {
  protected logMessage(msg: string): void {
    console.debug(`[DEBUG]: ${msg}`);
  }
}

class ErrorLogger extends Logger {
  protected logMessage(msg: string): void {
    console.error(`[ERROR]: ${msg}`);
  }
}

// Usage
enum LogTypes {
  INFO = "i",
  DEBUG = "d",
  ERROR = "e",
}

// Setup the chain of responsibility
const infoLogger = new InfoLogger(LogTypes.INFO);
const debugLogger = new DebugLogger(LogTypes.DEBUG);
const errorLogger = new ErrorLogger(LogTypes.ERROR);

// Linking handlers
infoLogger.setNext(debugLogger).setNext(errorLogger);

// Using the chain
infoLogger.handle(LogTypes.INFO, "This is an informational message");
infoLogger.handle(LogTypes.DEBUG, "This is a debug-level message");
infoLogger.handle(LogTypes.ERROR, "This is an error message.");

File: chain-of-responsibility-pattern/logger/logging-system.rs
trait Logger {
    fn set_next(&mut self, next: Box<dyn Logger>) -> &mut dyn Logger;
    fn log(&self, level: &str, message: &str);
}

struct BaseLogger {
    log_level: String,
    next_handler: Option<Box<dyn Logger>>,
}
impl BaseLogger {
    fn new(log_level: &str) -> Self {
        Self {
            log_level: log_level.to_string(),
            next_handler: None,
        }
    }
}

impl Logger for BaseLogger {
    fn set_next(&mut self, next: Box<dyn Logger>) -> &mut dyn Logger {
        self.next_handler = Some(next);
        // self.next_handler.as_mut().unwrap().as_mut()

        // as_deref_mut() converts Option<Box<T>> into Option<&mut T>
        self.next_handler.as_deref_mut().unwrap()
    }

    fn log(&self, level: &str, message: &str) {
        if self.log_level == level {
            println!("{}: {}", level, message);
        }

        if let Some(next_handler) = &self.next_handler {
            next_handler.log(level, message);
        }
    }
}

// Concrete loggers
// struct InfoLogger(BaseLogger);
// struct DebugLogger(BaseLogger);
// struct ErrorLogger(BaseLogger);

fn main() {
    let mut info_logger = Box::new(BaseLogger::new("INFO")) as Box<dyn Logger>;
    let debug_logger = Box::new(BaseLogger::new("DEBUG")) as Box<dyn Logger>;
    let error_logger = Box::new(BaseLogger::new("ERROR")) as Box<dyn Logger>;

    // Link the loggers
    info_logger.set_next(debug_logger).set_next(error_logger);

    // Using the chain of loggers
    info_logger.log("INFO", "This is an info message");
    info_logger.log("DEBUG", "This is a debug message");
    info_logger.log("ERROR", "This is an error message");
}
